### DISCUSSION TITLE
Usart Constants and enums - mixed deprecated

### DISCUSSION THREAD

**Discussion by mikrocoder** [SpenceKonde/DxCore#259] [Q&A]
Hi,

I will address the topic of UART constant again separately.

I often work with current avr-gcc toolchains. Many syntax is set to deprecated. For example that you should not mix different enums. Therefore I have to edit the UART_constant.h. It currently looks like this for me. This should not have a negative impact on other users. But it helps me a lot and others who use newer toolchains.

Example:
```cpp
In file included from C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\Arduino.h:26,
                 from C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART3.cpp:11:
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART.h: In member function 'virtual void UartClass::begin(long unsigned int)':
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART_constants.h:107:42: warning: bitwise operation between different enumeration types 'USART_CMODE_enum' and 'USART_SBMODE_enum' is deprecated [-Wdeprecated-enum-enum-conversion]
  107 |   #define SERIAL_8N1  (SERIAL_MODE_ASYNC | SERIAL_STOP_BIT_1 | SERIAL_PARITY_NONE | SERIAL_DATA_8)
      |                                          ^
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART.h:166:69: note: in expansion of macro 'SERIAL_8N1'
  166 |     void                   begin(unsigned long   baud) {begin(baud, SERIAL_8N1);}
      |                                                                     ^~~~~~~~~~
In file included from C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\Arduino.h:26,
                 from C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART0.cpp:29:
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART.h: In member function 'virtual void UartClass::begin(long unsigned int)':
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\cores\dxcore\UART_constants.h:107:42: warning: bitwise operation between different enumeration types 'USART_CMODE_enum' and 'USART_SBMODE_enum' is deprecated [-Wdeprecated-enum-enum-conversion]
  107 |   #define SERIAL_8N1  (SERIAL_MODE_ASYNC | SERIAL_STOP_BIT_1 | SERIAL_PARITY_NONE | SERIAL_DATA_8)
```

My modified [UART_constants.zip](https://github.com/SpenceKonde/DxCore/files/8180516/UART_constants.zip)

Now I wanted to ask if you would take over this?

**Comment by SpenceKonde** [comment]
Is this correct:
```
  #define SERIAL_DATA_5        (static_cast<uint8_t>(USART_CHSIZE_5BIT_gc & 0x04))
```
or should it be
```
  #define SERIAL_DATA_5        (static_cast<uint8_t>(USART_CHSIZE_5BIT_gc) & 0x04)
```

**Comment by SpenceKonde** [comment]
Also doesn't this mixed enum thing cause problems all over the place, not just here? I'm sure there are tons of places where group codes are OR'ed together.... because that's how the IO headers are meant to be used, and the version of the compiler that Microchip uses doesn't deprecate that. 

Can you check that SERIAL_DATA_5 matter and submit a PR to fix UART_constants.h?

**Comment by mikrocoder** [comment]
Hi,

I find the mixed enums only in DxCore. Do you have more mixed enums in DxCore? I know, because of your multi controller support.  :-)

Microchip has created any defines. Sometimes without understandable logic. MC also does not program C++ but only C and assembler.
The C++ standards are getting stricter and stricter to make programming more secure. Therefore many usual things were set to "deprecated". Here you have already set the toolchain to default C++17. Some small things are already more strict in the warnings. The big step comes when you start using C++20.

Because of
static_cast<uint8_t>(USART_CHSIZE_5BIT_gc & 0x04);
static_cast<uint8_t>(USART_CHSIZE_5BIT_gc) & 0x04;

It doesn't really matter. Only the first is a byte return value. The second is an int return value. I would use the first.
With the AVR DB both is 0. Because the 5BIT define is 0. I don't know if this has a value on other controllers. The syntax with static_cast is 100% C++ by the way. 

I'm still doing the PR ... [Link](https://github.com/SpenceKonde/DxCore/compare/master...mikrocoder:patch-1)

**Comment by SpenceKonde** [comment]
I think the numeric value of the CHSIZE bit is the same everywhere. (btw- the whole point of theat 0x04 is to change it to a reserved value. We later on test for that bit (and it doesn't actually get set),  and only if that bit is set do we treat it as 5-bit. That way, someone who omits the character size specification for whatever reason (likely ignorance) while specifying other options gets the behavior they were expecting (8-bit characters) not 5-bit ones, which probably only 0.01% of the Arduino community ever wants to use ;-) 

Yeah, I've noticed that Microchip's programming style isn't very sophisticated; that might be because their toolchain sucks at C++. Personally, I try to minimize use of C++ for one (well, two) reasons: The bad reason is that I'm not very good at it. The good reason, is that classes completely confound the optimizer, as do many other C++ features. Unreachable code ends up in binaries if it's a class method. Constants are not folded. LTO stands by and watches as incredibly inefficient code is produced which I'd expect to see optimized away to nothing.. I routinely see code in class methods where the comments claim that things are going to be optimized, you look at the assembly listing, and no optimization occurred.


I have at times been *extremely* tempted to use some regexes on the io headers, to turn the enums they supply into #defines    The regexes to do it wouldn't be hard.... It's more a matter of feeling that the I/O headers are sacred. 

If you wouldn't mind, can you check some of the examples for Event and Logic libraries? I know both use enumerated types, and they mat have similar issues.

**Comment by mikrocoder** [comment]
Hi,

the event examples do not show any problems. The Logic examples show problems. Here in the header defined defines are used directly which are deprecated. For the USART, you have added an intermediate layer and created your own definitions. I don't know yet what I would do here. Single register assignments or an intermediate layer. I have to think about it.

Something else caught my eye. Your enums contain errors. I use a AVR128DB64.
TCB2 is still missing?

```cpp
Logic:

Modulate:
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Modulate\Modulate.ino: In function 'void setup()':
Modulate:33:23: error: 'tcb1' is not a member of 'in'; did you mean 'tcb'?
   33 |   Logic0.input1 = in::tcb1;             // Use TCB1 WO as input 1
      |                       ^~~~
      |                       tcb
Bibliothek Logic in Version 1.1.2 im Ordner: C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic  wird verwendet
exit status 1
'tcb1' is not a member of 'in'; did you mean 'tcb'?


Oscillate:
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino: In function 'void demo9a()':
Oscillate:467:25: error: 'tca0_cnta' is not a member of 'user'; did you mean 'tca0_cnt_a'?
  467 |   Event1.set_user(user::tca0_cnta);         // Connect Event1 (carrying Logic0 output) to TCA0 event a
      |                         ^~~~~~~~~
      |                         tca0_cnt_a

Bibliothek Logic in Version 1.1.2 im Ordner: C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic  wird verwendet
Bibliothek Event in Version 1.2.0 im Ordner: C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Event  wird verwendet
exit status 1
'tca0_cnta' is not a member of 'user'; did you mean 'tca0_cnt_a'?
```

**Reply by SpenceKonde** [reply]
Oh my..... wait a minute, how could not be compiling? They pass the CI.... Looking into this now 

and here I was coming in here to give you the good news that your uart constants would be in the release.... And now code I was sure worked you say doesn't? I know I test compile logic on those parts..... 

```
Here in the header defined defines are used directly
```
I think this may be getting lost in translation. 
```
For the USART, you have added an intermediate layer and created your own definitions.
```
Well yeah how else to pack all those options into as few bytes as possible (notice how I leave them in the same bit positions as they end up in for the most part, so I can just use bitmasks! Compare with official Arduino core... which shuffles the bits around and then unshuffles them at runtime. I'm not making that up....
```
 I don't know yet what I would do here. Single register assignments or an intermediate layer. I have to think about it.
```
Still not sure where the issue is exactly.....

**Comment by mikrocoder** [comment]
Hello,

what is the error message with the new UART_constant.h? With which controller?

**About the Logic examples Modulate and Oscillate.** 
There is something wrong in your header files.

**Modulate:**
```cpp
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Modulate\Modulate.ino: In function 'void setup()':
Modulate:33:23: error: 'tcb1' is not a member of 'in'; did you mean 'tcb'?
   33 |   Logic0.input1 = in::tcb1;             // Use TCB1 WO as input 1
      |                       ^~~~
      |                       tcb
Bibliothek Logic in Version 1.1.2 im Ordner: C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic  wird verwendet
exit status 1
'tcb1' is not a member of 'in'; did you mean 'tcb'?
```

**Oscillate:**
```cpp
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino: In function 'void demo9a()':
Oscillate:467:25: error: 'tca0_cnta' is not a member of 'user'; did you mean 'tca0_cnt_a'?
  467 |   Event1.set_user(user::tca0_cnta);         // Connect Event1 (carrying Logic0 output) to TCA0 event a
      |                         ^~~~~~~~~
      |                         tca0_cnt_a

Bibliothek Logic in Version 1.1.2 im Ordner: C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic  wird verwendet
Bibliothek Event in Version 1.2.0 im Ordner: C:\Arduin IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Event  wird verwendet
exit status 1
'tca0_cnta' is not a member of 'user'; did you mean 'tca0_cnt_a'?
```

**About the enums in the Logic examples.**
Here we use directly the definitions from the controller header file. All different enums. I didn't show this because this is currently only my problem and there is no easy solution like with the usart.
```cpp
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino:474:48: warning: bitwise operation between different enumeration types 'TCA_SPLIT_CMD_enum' and 'TCA_SPLIT_CMDEN_enum' is deprecated [-Wdeprecated-enum-enum-conversion]
  474 |   TCA0.SPLIT.CTRLESET = TCA_SPLIT_CMD_RESET_gc | TCA_SPLIT_CMDEN_BOTH_gc;
      |                         ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino: In function 'void demo9d()':
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino:620:37: warning: bitwise operation between different enumeration types 'TCD_CNTPRES_enum' and 'TCD_SYNCPRES_enum' is deprecated [-Wdeprecated-enum-enum-conversion]
  620 |   TCD0.CTRLA = (TCD_CNTPRES_DIV4_gc | TCD_SYNCPRES_DIV8_gc | TCD_CLKSEL_OSCHF_gc);
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino: In function 'void demo10()':
C:\Arduino IDE Portable\AVR128DB\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.4.7\libraries\Logic\examples\Oscillate\Oscillate.ino:675:37: warning: bitwise operation between different enumeration types 'TCD_CNTPRES_enum' and 'TCD_SYNCPRES_enum' is deprecated [-Wdeprecated-enum-enum-conversion]
  675 |   TCD0.CTRLA = (TCD_CNTPRES_DIV1_gc | TCD_SYNCPRES_DIV1_gc | TCD_CLKSEL_PLL_gc);
      |                 ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
```


Either you do on user level as I did in UART_constant.h.
```cpp
TCA0.SPLIT.CTRLESET = static_cast<uint8_t>(TCA_SPLIT_CMD_RESET_gc) | static_cast<uint8_t>(TCA_SPLIT_CMDEN_BOTH_gc);
```
Looks unfamiliar stupid to many people, though. Correct according to C++ standard.
Or you can show it with single assignments.
Example:
```cpp
TCA0.SPLIT.CTRLESET = TCA_SPLIT_CMD_RESET_gc;
TCA0.SPLIT.CTRLESET |= TCA_SPLIT_CMDEN_BOTH_gc;
```

Or you create an additional layer with own name like UART_constant.h where you cast hidden.  Whether this really makes sense I don't know. According to my opinion you can't wrap the CCL in a lib anyway. There are too many possible combinations. In the examples you can see that despite the lib still much is done with direct register assignments. I generally program the CCL manually. You always have to read the manual.

But first we fix the problem with the UART_constant.h. How can I help? Did you change something in your toolchain? Used new controller header files? Microchip likes to change the bit names. I have often noticed this with the AVR DB.

I hope you can understand the translation.

**Comment by SpenceKonde** [comment]
Sorry for being unclear, usart_constants works. The unexpected compile error I referred to was the issue you found in the examples. 

And I got the examples compiling now (this was a problem that I didn't know existed because they're not in CI - misspelled a constant in oscillate, and used the tcb1 input in modulate (tcbs asCCL inputs aren't done like that on full size parts, only tinies - Logic was first ported to tinies, then from there to Dx, Event went the other direction - I can't blame MCUDude for not wanting to deal with the tinyAVR 0/1-series event system. It's a total shitshow, the designers clearly didn't know what they wanted until they'd gone too far down the wrong path to fix it)

And yeah, I usually use the CCL directly - but note that in the examples, the stuff being configured with direct register writes is the other peripherals being used as event sources/users! Every combination of options on the CCL is supported through the library. The CCL itself isn't terribly complicated. That it can connect to every other part of the chip and do all manner of 

--

So they don't like us ORing enums.... but that's how microchip intends us to use them. Microchip really should have just used defines for the group codes instead of fucking around with enums like that (then we could test for them in preprocessor statements too!), but as we know, they're not thinking about C++ - and we probably don't want them to, god knows what they'd screw up if they tried. 

I don't even know if splitting writes to CTRLESET will even work correctly; it might be order dependent - CTRLE is a really wacky register. TCAs and TCDs each have at least one weirdo register that has a set and clr register, and can autoclear itself when you issue a command - I don't fully understand those registers, the only thing I do with that register on a TCA is the hard reset command). It's clear that they had planned to allow for some commands to go to one channel at a time in split mode but then discovered it didn't work right and fixed it by removing all the options except NONE and BOTH.
If those registers work the way I think they are supposed to (which may or may not be how they work), it is never correct to write to _____SET (or _____CLR) register with |= or &= - the compiler doesn't know anything about those registers being special, and does it's usual read-modify-write, which generates extra unneeded code with the SET registers and unexpected behavior with the CLR registers. 

If the deprecated warnings are coming from code that is in the example, I think they stay, as the "correct" way is not deprecated on the version of C++ that we're compiling for and makes the code less readable, compromising it's usefulness as an example (and those examples are ugly already), and someone who has changed to C++ 20 doesn't need to use the examples as written.

If there are places where we do that in the library, I think using static casts is the appropriate fix. 
Changing to multiple lines with |= is no good - it adds at least 4 bytes and probably 6 to the compiled binary size and slows down execution, since the registers are all volatile. Remember that these changes get ported to megaTinyCore too - and we'll soon have 8k parts supported on DxCore too (for the EA series) 

If you didn't see, they released a first look at the headers for EA-series, so the EA is coming. Big changes to EVSYS, but the ADC looks like it's a direct copy of the 2-series tiny. And as with the 2-series,. they left in a chunk of header for internal use that's definitely not going to be there at release. 

I don't like the idea of putting an extra layer between the user and the headers - we had to keep that for USART for compatibility with Arduino (and we also had to pack the parameters into as few bits as possible while providing as many options as possible)

--

I hate the way Microchip has been changing bitfield names on us. Every time they make a breaking change, I have to add code to core_devices.h to make sure that either name works! I don't know *why* they are changing those names either. FREQSEL to FRQSEL? CRYSTAL to XTAL? 
I haven't rebuilt the toolchain with the newest atpacks in a while - that'll be done once we get indication that DD's are shipping and enable them all and have CI running against them, etc. I was expecting that to have happened months ago but....

**Comment by mikrocoder** [comment]
Hello,

okay, all is well.

My text should not be a reproach to you or MCUdude. You are doing a great job. Full respect. As always, it's all about very small subtleties.  :smile:

I also think you better leave it with the CCL as it is. Because of me nothing had to change. The enums I get myself under control.  :wink:

Because of the question about the SET and CLR registers. I think I can help there.
Example TCA - CTRLECLR / CTRLESET.
That is quite simply explained. It behaves like the PORT registers DIRCLR/DIRSET or OUTCLR/OUTSET.
I think this is a good feature because it makes the handling easier.
With this you don't have to clear bits yourself. The own read and clear (& ~) is omitted.
If you read these registers you always get 0 back anyway.
PORT.DIRCLR/DIRSET are higher-level of VPORT.DIR.
PORT.OUTCLR/OUTSET are higher-level registers of VPORT.OUT.
CLR/SET sets/clears only the desired bit finally in DIR or OUT register.
DIRSET ``` 0b 1111 0011 >>> DIR 0b 1111 0011```
DIRCLR ```0b 0110 0000 >>> DIR 0b 1001 0011```
You don't have to care about the logical connection of the bits yourself.

You can do it all yourself in the VPORT registers. Then you have to take care of all bits yourself like before. So read and clear them with & ~ etc. The SET/CLR registers do this work for you.

It is the same with TCA.CTRLCLR/CTRLSET.
A logical '1' in CTRLESET sets this bit.
A logical '1' in CTRLECLR clears exactly this bit.
Only that there is no "pure" CTLRE register for direct bit manipulation.

Realization: In the SET/CRL registers you only have to use OR to manipulate single bits. You only think positive about the bit you want to change. The SET or CLR register then determines whether this bit is set or cleared. You get used to it very fast.

The bit name renaming in the header files also annoys me. I wanted to ask Microchip already once what that should. Whether they can agree finally once. But I never did. 

 I would have a crazy idea because of the enums. You write your own headerfile. This time sorted by registers. All bit definitions of a register belong to an enum. Makes also logical sense. Makes a lot of work and must not contain a single error. It's just a crazy idea. :wink:
I don't know the AVR EA series yet. I concentrate first on the DB series. This will be my development platform.

Translated with www.DeepL.com/Translator (free version)

**Reply by SpenceKonde** [reply]
Don't think I haven't considered replacing the header files with my own versions. But I think I need to treat the headers as sacred, because I know too many people would flip their shit if I tried something like that.

At one point I also considered duplicating every PERIPERAL_OPTION_ONE_gc  groupcode as a #define PERIPHERAL_OPTION_ONE_GC equal to the same value, and putting that in something included by core_devices.h - generating it wouldn't be that hard - the names are systematic enough it's an easy regex..... 
(assuming sublime text, butother editors have similar features, my point is it's mot hard)
Just search
```
^ +[A-Z0-9_]*gc = \(.*?\)
```
find-all, ctrl+c ctrl+n ctrl+v 
then find/replace 
```
^ +([A-Z0-9_]*)gc = (\(.*?\))
```
with
```
#define \1GC \2
```


But I just don't see this as something that is a priority to do until we're forced to move to C++20. And each version of C++ as seemed less and less appropriate for microcontrollers, with more and more baggage that we have to deal with, but doesn't get used in any way, so I don't see any reason to rush to C++ 20. In the case of USART constant's there was an easy fix for the few users who are on C++ 20 for whatever reason, for tings that require development/maintenance effort, I think the sleep+power save library and a few dozen other things are a higher priority 

Until last year the core compiled to C++11,  and the official one still does!

Oh I am very well aware of how the set and clear registers work on the ports. I just don't understand why they had to do something like that in those timer registers, and I was under the impression that at least some of those bits would auto-clear when the command was executed, further complicating thing. 
They carry a huge cost in terms of the developer confusion they cause when use |= or &= to assign to a clr register. I've gotten so many support requests from people who appear to have spent days trying to figure out what was going on when they were just doing 
```
PORTA.DIRCLR |= 1<<4 ; // this will clear every bit in the register
```
instead of
```
PORTA.DIRCLR = 1<<4; //this will clear bit 4. 
```

My point was more that I haven't looked closely enough to be sure that the CTRLESET/CLR and CTRLFSET/CLR registers act quite as you would expect looking at the port registers, or if there are added complications to them.
